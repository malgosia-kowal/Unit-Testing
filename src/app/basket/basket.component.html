<div class="basketContainer">
  <div class="basketContainerDetail">
    <div> Total: {{ basketService.total }} zl</div>
    <div> Products: {{ basketService.products.length }}</div>
  </div>
  <div>
    <app-button
      text="Clean"
      (action)="basketService.clear()"
    >
    </app-button>
  </div>
</div>
