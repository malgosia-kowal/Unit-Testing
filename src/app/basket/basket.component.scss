.basketContainer {
  z-index: 999;
  position: fixed;
  bottom: 0;
  background: white;
  width: 100vw;
  box-shadow: 0 0 10px rgb(0, 0, 0, 0.3);
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: 10px;

  &Detail {
    display: flex;
    flex-direction: column;
    justify-content: space-evenly;

    & > div {
      font-size: 20px;
    }
  }

  &Buttons {
    display: flex;
    flex-direction: row;
    margin-right: 40px;
  }
}
