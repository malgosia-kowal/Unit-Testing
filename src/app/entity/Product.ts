export class Product {
  public name: string;
  public price: number;
  public quantity: number;
  public size: string;
  public image: string;

  constructor(name: string, price: number, quantity: number, size: string, image: string) {
    this.name = name;
    this.price = price;
    this.quantity = quantity;
    this.size = size;
    this.image = image; 
  }
}
