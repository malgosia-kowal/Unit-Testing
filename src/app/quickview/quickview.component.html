<div *ngIf="visible">
  <div (click)="toggle()" class="overlay"></div>
  <div class="quickviewContainer">
    <h4>{{ "QuickView.Header" | translate }}</h4>
    <div *ngIf="!products.length">{{ "QuickView.Empty" | translate }}</div>
    <div>
      <div class="quickviewProduct" *ngFor="let product of products">
        <div>
          <img src="{{ product.image }}" />
        </div>
        <div>
          <div>{{ "QuickView.Name" | translate: product }}</div>
          <div>{{ "QuickView.Color" | translate: product }}</div>
          <div>
            {{ "QuickView.Price" | translate: product }}
            {{ "Currency" | translate }}
          </div>
          <div>{{ "QuickView.Quantity" | translate: product }}</div>
        </div>
        <div
          class="closeIcon"
          (click)="basketService.removeProduct(product.name, product.size)"
        ></div>
      </div>
    </div>
  </div>
</div>
