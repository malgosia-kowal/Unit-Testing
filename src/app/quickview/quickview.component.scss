.quickviewContainer {
  position: fixed;
  right: 0;
  top: 0;
  height: 100%;
  background: white;
  box-shadow: 0 0 10px rgb(0, 0, 0, 0.3);
  padding: 10px;
}

.quickviewProduct {
  display: flex;
  flex-direction: row;
  padding: 5px 0;

  & > div:first-of-type {
    height: 75px;
    & img {
      object-fit: contain;
      height: 100%;
    }
  }

  & > div:nth-of-type(2) {
    padding: 0 5px;
    display: flex;
    flex-direction: column;
  }
}
