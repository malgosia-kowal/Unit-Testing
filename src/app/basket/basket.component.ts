import { Component, OnInit } from '@angular/core';
import { BasketService } from '../service/basket.service';
import { Toggable, ToggleService } from '../service/toggle.service';

@Component({
  selector: 'app-basket',
  templateUrl: './basket.component.html',
  styleUrls: ['./basket.component.scss']
})
export class BasketComponent implements OnInit {

  constructor(public basketService: BasketService, public toggleService: ToggleService) { }

  ngOnInit() {
  }

  toggleQuickview() {
    this.toggleService.toggle(Toggable.Quickview);
  }

}
