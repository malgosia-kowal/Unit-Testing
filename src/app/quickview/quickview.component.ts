import { Component, OnInit } from '@angular/core';
import { Toggable, ToggleService } from '../service/toggle.service';
import { Product } from '../products/product';
import { BasketService } from '../service/basket.service';

@Component({
  selector: 'app-quickview',
  templateUrl: './quickview.component.html',
  styleUrls: ['./quickview.component.scss']
})
export class QuickviewComponent implements OnInit {
  visible: boolean;
  products: Product[];
  constructor(public toggleService: ToggleService, private basketService: BasketService) { }

  ngOnInit() {
    this.toggleService.visible(Toggable.Quickview).subscribe(visible => {
      this.visible = visible;
    });
    this.products = this.basketService.products;
  }

}
