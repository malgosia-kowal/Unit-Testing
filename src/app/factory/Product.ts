import { Product } from '../entity/Product'; 
import * as faker from 'faker';

export const createProduct = (data: Partial<Product> = {}) => new Product(
    data.hasOwnProperty('name') ? data.name : faker.commerce.productName(), 
    data.hasOwnProperty('price') ? data.price : parseInt(faker.commerce.price()), 
    1, 
    data.hasOwnProperty('size') ? data.size : faker.commerce.color(),
    data.hasOwnProperty('image') ? data.image : faker.image.imageUrl(),
);
