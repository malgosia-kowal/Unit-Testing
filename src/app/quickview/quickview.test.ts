  import {
    async,
    ComponentFixture,
    TestBed,
  } from '@angular/core/testing';
  import { By } from '@angular/platform-browser';
  import { BasketService } from '../service/basket.service';
  import { ToggleService, Toggable } from '../service/toggle.service';
  import { createProduct } from '../factory/Product';
  import {QuickviewComponent} from './quickview.component';
  import { BehaviorSubject } from 'rxjs';
  
  describe('Quickview component', () => {
    let quickviewComponent: QuickviewComponent;
    let fixture: ComponentFixture<QuickviewComponent>;
    let toggleService: ToggleService;
    let basketService: BasketService;
    
    beforeEach(async(async () => {
      TestBed.configureTestingModule({
        declarations: [QuickviewComponent],
        providers: [
          { provide: ToggleService, useClass: ToggleService },
          { provide: BasketService, useClass: BasketService }
        ]
      })
        .compileComponents()
        .then(() => {
          fixture = TestBed.createComponent(QuickviewComponent);
          quickviewComponent = fixture.componentInstance;

          toggleService = TestBed.get(ToggleService);
          basketService = TestBed.get(BasketService);

          fixture.detectChanges();
        });
    }));
    it('should exist', () => {
      expect(quickviewComponent).toBeDefined();
    });
  
    it('should render closed by default', () => {
      expect(fixture).toMatchSnapshot();
    });
  
    it('can render empty', () => {
      quickviewComponent.toggleService.toggle(Toggable.Quickview);

      expect(quickviewComponent.visible).toBeTruthy();

      fixture.detectChanges();

      expect(fixture).toMatchSnapshot();
    });
  
    it('should set the dependecies', () => {
      quickviewComponent.ngOnInit();
  
      expect(quickviewComponent.toggleService).toBe(toggleService);
    });
  
    it('should render products when is visible', () => {
      const product = createProduct({ price:300, size:'green', name:'kity' });
      
      basketService.addProduct(product);

      quickviewComponent.toggleService.toggle(Toggable.Quickview);

      fixture.detectChanges();
      
      expect(basketService.products.getValue().length).toEqual(1);

      expect(fixture).toMatchSnapshot();
    });
  
    it('should remove product from quickView', () => {
      const product = createProduct();
      basketService.addProduct(product);
      
      quickviewComponent.toggleService.toggle(Toggable.Quickview);

      fixture.detectChanges();
  
      const removeProductButton = fixture.debugElement.query(By.css('.closeIcon'));
      removeProductButton.triggerEventHandler('click', 'test');

      fixture.detectChanges();

      expect(basketService.get().products.getValue().length).toEqual(0);
    });

    it('should close quickView when click on overlay', () => {
      quickviewComponent.toggleService.toggle(Toggable.Quickview);

      fixture.detectChanges();
  
      const overlay = fixture.debugElement.query(By.css('.overlay'));
      overlay.triggerEventHandler('click', null);
 
      const quickViewVisibleSubject= (toggleService.visible(Toggable.Quickview).source as BehaviorSubject<boolean>);
      
      expect(quickViewVisibleSubject.getValue()).toBeFalsy();

      fixture.detectChanges();
      
      expect(fixture).toMatchSnapshot();
    });
  });