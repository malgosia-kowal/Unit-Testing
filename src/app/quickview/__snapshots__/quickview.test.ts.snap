// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quickview component can render empty 1`] = `
<app-quickview
  basketService={[Function BasketService]}
  products={[Function Array]}
  toggleService={[Function ToggleService]}
  visible={[Function Boolean]}
>
  <div>
    <div
      class="overlay"
    />
    <div
      class="quickviewContainer"
    >
      <h4>
        Your basket
      </h4>
      
      <div>
        No Cats in the Basket
      </div>
      <div>
        
      </div>
    </div>
  </div>
</app-quickview>
`;

exports[`Quickview component should close quickView when click on overlay 1`] = `
<app-quickview
  basketService={[Function BasketService]}
  products={[Function Array]}
  toggleService={[Function ToggleService]}
  visible="false"
>
  
</app-quickview>
`;

exports[`Quickview component should render closed by default 1`] = `
<app-quickview
  basketService={[Function BasketService]}
  products={[Function Array]}
  toggleService={[Function ToggleService]}
  visible="false"
>
  
</app-quickview>
`;

exports[`Quickview component should render products when is visible 1`] = `
<app-quickview
  basketService={[Function BasketService]}
  products={[Function Array]}
  toggleService={[Function ToggleService]}
  visible={[Function Boolean]}
>
  <div>
    <div
      class="overlay"
    />
    <div
      class="quickviewContainer"
    >
      <h4>
        Your basket
      </h4>
      
      <div>
        
        <div
          class="quickviewProduct"
        >
          <div>
            <img
              src="http://lorempixel.com/640/480"
            />
          </div>
          <div>
            <div>
              Name: kity
            </div>
            <div>
              Color: green
            </div>
            <div>
              Price: 300 zl
            </div>
            <div>
              Quantity: 1
            </div>
          </div>
          <div
            class="closeIcon"
          />
        </div>
      </div>
    </div>
  </div>
</app-quickview>
`;
